File: python/src/anysearch/providers/google_pse.py
"""Google Programmable Search Engine (Custom Search JSON API)."""

from __future__ import annotations

import os
from typing import Any, Dict, Optional

import httpx

from .._http import PreparedRequest
from ..types import SearchRequest, SearchResponse
from .base import BaseProvider, Capability

_CX_ENV = ("GOOGLE_PSE_CX", "GOOGLE_CSE_CX", "GOOGLE_PSE_ENGINE_ID")


class GooglePSEProvider(BaseProvider):
    name = "google_pse"
    aliases = ("google", "google_cse")
    env_keys = ("GOOGLE_PSE_API_KEY", "GOOGLE_CSE_API_KEY", "GOOGLE_API_KEY")
    default_base_url = "https://www.googleapis.com"
    capabilities = frozenset(
        {
            Capability.COUNTRY,
            Capability.LANGUAGE,
            Capability.SAFE_SEARCH,
            Capability.DOMAINS,
        }
    )

    def __init__(self, api_key: Optional[str] = None, base_url: Optional[str] = None, **config: Any):
        super().__init__(api_key=api_key, base_url=base_url, **config)
        self.cx = config.get("cx") or self._cx_from_env(self._env)

    @staticmethod
    def _cx_from_env(env: Optional[Dict[str, str]] = None) -> Optional[str]:
        if env is None:
            env = os.environ  # type: ignore[assignment]
        for key in _CX_ENV:
            if env.get(key):
                return env[key]
        return None

    @classmethod
    def is_configured(cls, env: Optional[Dict[str, str]] = None) -> bool:
        if env is None:
            env = os.environ  # type: ignore[assignment]
        return bool(cls.key_from_env(env)) and bool(cls._cx_from_env(env))

    @classmethod
    def config_hint(cls) -> str:
        return "Set GOOGLE_PSE_API_KEY and GOOGLE_PSE_CX (search engine id)."

    def prepare(self, req: SearchRequest) -> PreparedRequest:
        params: Dict[str, Any] = {
            "key": self.api_key,
            "cx": self.cx,
            "q": req.query,
            "num": min(req.max_results, 10),
        }
        if req.country:
            params["gl"] = req.country.lower()
        if req.language:
            params["hl"] = req.language
            params["lr"] = f"lang_{req.language[:2]}"
        if req.safe_search:
            params["safe"] = "off" if req.safe_search == "off" else "active"
        if req.start_published_date or req.end_published_date:
            start = (req.start_published_date or "")[:10].replace("-", "")
            end = (req.end_published_date or "")[:10].replace("-", "")
            if start and end:
                params["sort"] = f"date:r:{start}:{end}"
        if req.include_domains:
            params["siteSearch"] = req.include_domains[0]
            params["siteSearchFilter"] = "i"
        elif req.exclude_domains:
            params["siteSearch"] = req.exclude_domains[0]
            params["siteSearchFilter"] = "e"
        params.update(req.extra)
        return PreparedRequest(
            method="GET",
            url=f"{self.base_url}/customsearch/v1",
            headers={"Accept": "application/json"},
            params=params,
        )

    def parse(self, response: httpx.Response, req: SearchRequest, elapsed_ms: float) -> SearchResponse:
        data = response.json()
        results = []
        for item in data.get("items", []) or []:
            results.append(
                self._result(
                    title=item.get("title"),
                    url=item.get("link"),
                    snippet=item.get("snippet"),
                    source=item.get("displayLink"),
                    raw=item,
                )
            )
        total = None
        info = data.get("searchInformation") or {}
        if info.get("totalResults") is not None:
            try:
                total = int(info["totalResults"])
            except (TypeError, ValueError):
                total = None
        return self._response(req, results, data, total_results=total, elapsed_ms=elapsed_ms)

File: js/src/providers/googlePse.ts
import type { HttpResponse } from "../http.js";
import type { Env, SearchRequest, SearchResponse } from "../types.js";
import { finalize, result, type ProviderContext, type ProviderSpec } from "./base.js";

const KEY_ENV = ["GOOGLE_PSE_API_KEY", "GOOGLE_CSE_API_KEY", "GOOGLE_API_KEY"];
const CX_ENV = ["GOOGLE_PSE_CX", "GOOGLE_CSE_CX", "GOOGLE_PSE_ENGINE_ID"];

function cxFromEnv(env: Env): string | undefined {
  for (const k of CX_ENV) if (env[k]) return env[k];
  return undefined;
}

export const googlePse: ProviderSpec = {
  name: "google_pse",
  aliases: ["google", "google_cse"],
  envKeys: KEY_ENV,
  defaultBaseUrl: "https://www.googleapis.com",
  capabilities: new Set(["country", "language", "safe_search", "domains"]),

  isConfigured(env: Env): boolean {
    const hasKey = KEY_ENV.some((k) => env[k]);
    return Boolean(hasKey && cxFromEnv(env));
  },
  configHint(): string {
    return "Set GOOGLE_PSE_API_KEY and GOOGLE_PSE_CX (search engine id).";
  },

  prepare(ctx: ProviderContext, req: SearchRequest) {
    const cx = (ctx.config.cx as string) ?? cxFromEnv(ctx.env);
    const params: Record<string, string | number | boolean | undefined> = {
      key: ctx.apiKey,
      cx,
      q: req.query,
      num: Math.min(req.maxResults, 10),
    };
    if (req.country) params.gl = req.country.toLowerCase();
    if (req.language) {
      params.hl = req.language;
      params.lr = `lang_${req.language.slice(0, 2)}`;
    }
    if (req.safeSearch) params.safe = req.safeSearch === "off" ? "off" : "active";
    if (req.startPublishedDate && req.endPublishedDate) {
      const s = req.startPublishedDate.slice(0, 10).replace(/-/g, "");
      const e = req.endPublishedDate.slice(0, 10).replace(/-/g, "");
      params.sort = `date:r:${s}:${e}`;
    }
    if (req.includeDomains.length) {
      params.siteSearch = req.includeDomains[0];
      params.siteSearchFilter = "i";
    } else if (req.excludeDomains.length) {
      params.siteSearch = req.excludeDomains[0];
      params.siteSearchFilter = "e";
    }
    for (const [k, v] of Object.entries(req.extra)) params[k] = v as string;
    return {
      method: "GET",
      url: `${ctx.baseUrl}/customsearch/v1`,
      headers: { Accept: "application/json" },
      params,
    };
  },

  parse(_ctx, res: HttpResponse, req: SearchRequest, elapsedMs: number): SearchResponse {
    const data = res.data ?? {};
    const results = (data.items ?? []).map((item: any) =>
      result({
        title: item.title,
        url: item.link,
        snippet: item.snippet,
        source: item.displayLink,
        raw: item,
      }),
    );
    const total = Number.parseInt(data.searchInformation?.totalResults ?? "", 10);
    return finalize(googlePse, req, results, data, {
      totalResults: Number.isNaN(total) ? undefined : total,
      elapsedMs,
    });
  },
};
